modules/data-streams/src/main/java/org/elasticsearch/datastreams/lifecycle/downsampling/ReplaceSourceWithDownsampleIndexTask.java

```diff
@@ -0,0 +1,205 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.datastreams.lifecycle.downsampling;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.ClusterStateTaskListener;
+import org.elasticsearch.cluster.metadata.DataStream;
+import org.elasticsearch.cluster.metadata.IndexAbstraction;
+import org.elasticsearch.cluster.metadata.IndexMetadata;
+import org.elasticsearch.cluster.metadata.Metadata;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.core.TimeValue;
+import org.elasticsearch.index.IndexSettings;
+
+import java.util.HashMap;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Objects;
+
+import static org.elasticsearch.datastreams.DataStreamsPlugin.LIFECYCLE_CUSTOM_INDEX_METADATA_KEY;
+
+/**
+ * Cluster state task that replaces a source index in a data stream with its downsample index.
+ * In the process it will configure the origination date for the downsample index (so it can
+ * have a correct generation time).
+ */
+public class ReplaceSourceWithDownsampleIndexTask implements ClusterStateTaskListener {
```

Comment:
We're adding an executor and task in this PR, but I wonder if instead it would be better for us to teach the `MetadataDataStreamsService` about this, and then make use of that method (we already have `addBackingIndex` and `removeBackingIndex` as static methods there, we could add `replaceBackingIndex`) so that all the logic still lives inside of that class. What do you think?

Reply:
(We could potentially add a passed-in function to `replaceBackingIndex` that would allow copying the origination date piece or anything else)